### PR TITLE
Allow for Top-Level References

### PR DESCRIPTION
**Pull request by erikerikson**
## What did you implement:

Closes https://github.com/serverless/serverless/issues/3207 
Closes https://github.com/serverless/serverless/issues/3206

## How did you implement it:

1. Newly allow `${self:}` as a valid reference to the current serverless.yml's root
  a. This allows for a more clean manner of passing names and identifiers around between projects.  Particularly, this facilitates exporting values from projects so they can be used in other projects via Fn::ImportValue.
2. Newly allow `${opt:}` and `${env:}` as a valid reference to the root of the option and environment settings hive, respectively
3. Write a test ensuring 1. and 2. above
3. Document the new capabilities
4. Provide explicit support for the use of populateProperty in a not-in-place manner

Example usage of `${self:}`: https://github.com/Nordstrom/hello-retail/pull/26/files

## How can we verify it:

using hello-retail without the private.yml and some backing implementation will be a bit rough.  I'm happy to discuss if use isn't clear.  The associated issue may help a little.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [ ] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO (additive only)

### CODE REVIEW

File: lib/classes/Variables.js [hunk 1]
```diff
@@ -145,7 +148,7 @@ class Variables {
   }
 
   getValueFromSelf(variableString) {
-    let valueToPopulate = _.cloneDeep(this.service);
+    let valueToPopulate = this.service;
```
Comment (eahefnawy) [inline lib/classes/Variables.js]: we clone just in case because objects are passed by reference, so any changes to `valueToPopulate` would actually affect the original service object. Is there any reason you removed cloning?

File: lib/classes/Variables.js [hunk 2]
```diff
@@ -216,7 +219,7 @@ class Variables {
     deepProperties.forEach(subProperty => {
       if (typeof valueToPopulate === 'undefined') {
         valueToPopulate = {};
-      } else {
+      } else if (subProperty !== '') {
```
Comment (eahefnawy) [inline lib/classes/Variables.js]: could you clarify this extra conditional? Why is it necessary?

File: lib/classes/Variables.js (hunk 1 above)
Comment (erikerikson) [inline lib/classes/Variables.js]: I thought this was the whole point!  ;)

Actually, changing this moved me closer to enabling my scenario but may be legacy no longer necessary because of other changes.  I have to admit that I haven't completely groked this code.

What I observed was that the attribute value ${self:} was being copied and then populate was being called again.  I was concerned that it would infinitely recurse on that.  I saw evidence later that expansion is on demand so I will try and remove this today and test it out some more.

File: lib/classes/Variables.js (hunk 2 above)
Comment (erikerikson) [inline lib/classes/Variables.js]: You bet.  The additional conditional will resolve to true in the case of ${self:}, any multiple dots in a row (e.g. ${[...]attr1..attrNext[...]}), or a trailing dot (e.g. ${[...]attr1.attrNext.}).  In each case, '""' in valueToPropagate will evaluate to false (I think? Actually, maybe I should add that check because I think obj[''] = {} is valid code, technically speaking) and the empty string indicates that the current value of valueToPropagate should continue to be the current value.  Another way to state this is that the empty string does not represent an attribute to traverse, therefore no traversal should occur and the current state continues to be the specified valid state.

Apologies for lack of formatting and whatnot, on the phone this morning and don't have a grave'.

Comment (erikerikson) [inline lib/classes/Variables.js]: I'll expand this conditional today too

File: lib/classes/Variables.js (hunk 1 above)
Comment (erikerikson) [inline lib/classes/Variables.js]: [sorry for the novel...]

TL;DR: 1) avoid infinite recursion 2) serverless depends on in-place affect on the original service object.

Detailed Discussion:

So, my reasoning was at least correctish...  Not sure that my "evidence ... [of] expansion on demand" was properly interpreted.

Re-adding the `_.deepClone` causes an infinite recursion (technically, it only got 177 iterations deep before I paused execution).  It copies the `custom.retailStream: ${self:}` attribute and value pair that is to be replaced and then (because `if (property !== this.service)` fails to identify the property is a copy of the service [on line 73]) attempts to `populateProperty` on the service copy (that still contains the `${self:}`.  This repeats.  That's layer one of the infinite recursion.

I could write a rough equivalency check to avoid the recursion on line 73.  Something like `if (!(property.service && property.service === this.service.service))` would probably suffice in most cases but `_.deepEquals(property, this.service)` definitely wouldn't work since the two will have already diverged (the parent service has an extra level that the cloned service does not).

The rough equivalency check will get us past the recursion on line 73 but then the traverse on line 41 will attempt to traverse the `${self:}` and as a result get another one, et cetera, to "infinity and beyond!". 😄 

It seems to me that if we want to protect the "original" in-memory copy of the service then we ought to do that once at the outset (i.e. in the constructor or the like).  The problem is that this doesn't cover the bare use of `populateProperty(<something-not-the-service>)` but a revision to the call that adds a "called from populateService" parameter could distinguish between the cases.  Passing the property in the `populateProperty` case but (relying on instance state and) not passing the service in the `populateService` case is a bit of a mental model mix-up that is legacy now.

Hmm... (sorry, writing as I code and test...)

It turns out that the framework is actually **relying** on the revision of the original service object.  Adding the use of `_.cloneDeep` to the constructor causes the copy of the service that the framework uses to take action not on the altered copy but the original and thus I observed:
```
  Serverless Error ---------------------------------------
 
     Bucket name must start with a letter or number. ${self:custom.helloRetail.deploymentBucket}
 
  Stack Trace --------------------------------------------
 
ServerlessError: Bucket name must start with a letter or number. ${self:custom.helloRetail.deploymentBucket}
    at ~/serverless/lib/plugins/aws/lib/validate.js:57:17
From previous event:
    at AwsDeploy.validateS3BucketName (~/serverless/lib/plugins/aws/lib/validate.js:34:8)
    at AwsDeploy.<anonymous> (~/serverless/lib/plugins/aws/deploy/lib/configureStack.js:22:26)
From previous event:
    at AwsDeploy.configureStack (~/serverless/lib/plugins/aws/deploy/lib/configureStack.js:22:10)
From previous event:
    at __dirname.constructor.options.hooks.deploy:initialize (~/serverless/lib/plugins/aws/deploy/index.js:42:10)
    at ~/serverless/lib/classes/PluginManager.js:160:50
    at processImmediate [as _immediateCallback] (timers.js:383:17)
From previous event:
    at PluginManager.run (~/serverless/lib/classes/PluginManager.js:160:22)
    at Serverless.run (~/serverless/lib/Serverless.js:91:31)
    at ~/serverless/bin/serverless:23:50
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
 
  Your Environment Information -----------------------------
     OS:                 darwin
     Node Version:       4.6.2
     Serverless Version: 1.5.1
```

So therefore... I can remove the top level clone and still do a clone below if the caller doesn't ask for an in-place population (which populate service would ask for).  The remaining potential concern space then is that the service could be improperly altered by the population of a property, via a `${self:<var>*}` declaration that isn't a part of the service.  However, by definition, the `self` variable type only references the service, so that concern space is empty/size zero.

Maybe I fail to understand the case that you were concerned about that motivated the cloning?  Sorry if so.  I'd love some illumination.

**Comment by erikerikson** [comment]
Added fix for https://github.com/serverless/serverless/issues/3206 as well.  Not that it only provides a warning, not an error, so as not to create a breaking change.

**Comment by erikerikson** [comment]
As mentioned briefly in #3206, cross-serverless-service loading causes the kinds of errors that are warned about.  This potentially means that my fix here and strategy for allowing consistent cross-project resolution is unwise.  A potentially better fix is to resolve all `self` variables to the file from whence they came and add a `local` variable type for rendering using the context into which you are being loaded.  The loading of an external file would thus be able to resolve in its own or the loader's context as expected/intended by the file creator.

File: docs/providers/aws/guide/variables.md [hunk 3]
```diff
@@ -41,12 +45,18 @@ functions:
...
+      Value: 'A Value To Export'
+      Export:
+        Name: ${self:custom.exportName}
 ```
 
 In the above example you're setting a global schedule for all functions by referencing the `globalSchedule` property in the same `serverless.yml` file. This way, you can easily change the schedule for all functions whenever you like.
 
 ## Referencing Environment Variables
-To reference environment variables, use the `${env:SOME_VAR}` syntax in your `serverless.yml` configuration file.
+To reference environment variables, use the `${env:SOME_VAR}` syntax in your `serverless.yml` configuration file.  `${env:}` references the environment variable hive.
```
Comment (pmuens) [inline docs/providers/aws/guide/variables.md]: Could we explain that a little bit 😬 . It's hard to understand what this means for me as a non-native speaker (my translator says smth. about  🐝 🐝 🐝 )

File: docs/providers/aws/guide/variables.md [hunk 4]
```diff
@@ -63,7 +73,7 @@ functions:
 In the above example you're dynamically adding a prefix to the function names by referencing the `FUNC_PREFIX` env var. So you can easily change that prefix for all functions by changing the `FUNC_PREFIX` env var.
 
 ## Referencing CLI Options
-To reference CLI options that you passed, use the `${opt:some_option}` syntax in your `serverless.yml` configuration file.
+To reference CLI options that you passed, use the `${opt:some_option}` syntax in your `serverless.yml` configuration file.    `${opt:}` references the options hive.
```
Comment (pmuens) [inline docs/providers/aws/guide/variables.md]: Same comment about explanation like above.

File: lib/classes/Variables.js [hunk 5]
```diff
@@ -130,20 +141,52 @@ class Variables {
     return valueToPopulate;
   }
 
+  warnIfNotFound(variableString, valueToPopulate) {
```
Comment (pmuens) [inline lib/classes/Variables.js]: Can we add a unit test for this?

File: lib/classes/Error.js [hunk 6]
```diff
@@ -2,6 +2,40 @@
 const chalk = require('chalk');
 const version = require('./../../package.json').version;
 
+const consoleLog = (message) => {
+  console.log(message); // eslint-disable-line no-console
+};
+
+const writeMessage = (messageType, message) => {
```
Comment (pmuens) [inline lib/classes/Error.js]: Can we add a unit test for this?

File: lib/classes/Error.js [hunk 7]
```diff
@@ -90,3 +95,7 @@ module.exports.logError = (e) => {
     throw new Error(e);
   }
 };
+
+module.exports.logWarning = (message) => {
```
Comment (pmuens) [inline lib/classes/Error.js]: Can we add a unit test for that?

**Comment by erikerikson** [comment]
There is an ugly consequence of not having the `local` reference type.  While I have successfully included the exports of other services, the dependencies that must be declared are cumulative through dependency chains.

Concretely:
1. Service A
2. Service B
  a. B depends on A, using one of its exports.  It must declare `custom.A` and `custom.AVersion`.
3. Service C
  a. C depends on B, using one of its exports.  In order to load depend on B successfully, it has to declare `custom.B` and `custom.BVersion` but also `custom.A` and `custom.AVersion`.
  b. custom.AVersion must match what the version of B declared for custom.AVersion or obscure hilarity will ensue.

File: docs/providers/aws/guide/variables.md (hunk 3 above)
Comment (erikerikson) [inline docs/providers/aws/guide/variables.md]: Of course.  Thanks for calling out the poor communication!  👍 
I'm thinking that I'll just be concrete and reference `process.env`.

File: docs/providers/aws/guide/variables.md (hunk 4 above)
Comment (erikerikson) [inline docs/providers/aws/guide/variables.md]: 👍

**Comment by erikerikson** [comment]
However, the `local` bit probably deserves a separate Issue and PR.

File: docs/providers/aws/guide/variables.md (hunk 3 above)
Comment (pmuens) [inline docs/providers/aws/guide/variables.md]: Thanks @erikerikson!

No hard feelings. A reference to `process.env` sounds good! 👍

**Comment by erikerikson** [comment]
It could be a bit too paranoid but worth at least a moment's consideration...  For the `${env:}` scope of this change, there is a security risk being created.  The warning I have added to the docs (it'll be checked in shortly) follows.  This may influence whether we want that to be in scope for this PR.

```
!!!WARNING!!! If you use use the environment to provide sensitive data such as credentials or private keys to your project, there is a risk that this sensitive information could be written into less protected or publicly accessible build logs, CloudFormation Templates, et cetera. !!!WARNING!!!
```

I don't believe there are any sensitive items expected to be in the sls options.  If this is incorrect, we should definitely discuss the concern for that scope as well.

File: lib/classes/Error.js (hunk 6 above)
Comment (erikerikson) [inline lib/classes/Error.js]: writeMessage is an unchanged movement of code from the logError method to a private method of the module.  Because it is a private method, it is not externally available.  Given its use by logError, it gets the same exact coverage as before plus overlapping coverage from the new `logWarning` tests.

**Comment by erikerikson** [comment]
@pmuens please...
1. let me know if the documentation clarifications could still use improvement.
2. feel free to suggest different ways to think on the `Error.js/writeMessage` tests topic.  I think my considerations are correct but feel free to contradict.
3. let me know what you (and invite anyone else who should weigh in) think of the "WARNING" comment.
4. mention anything else that should occur, of course

**Comment by pmuens** [comment]
Thanks for the updates @erikerikson added it to my todo list 👍

File: lib/classes/Error.js (hunk 6 above)
Comment (pmuens) [inline lib/classes/Error.js]: Yep, you're totally right @erikerikson 👍

File: lib/classes/Variables.js (hunk 1 above)
Comment (pmuens) [inline lib/classes/Variables.js]: @eahefnawy for me it makes sense to get rid of deep cloning to prevent an infinite loop / recursion.

What's your take on this since you've implemented it in the first place?

Comment (eahefnawy) [inline lib/classes/Variables.js]: @erikerikson @pmuens I think you guys are right. I've been doing this since v0.5 because in v0.5 we used to serialize `serverless.yml`, so we wanted to keep an immutable version of the state of the file as it was before loading, hence the cloning. We don't do that anymore in v1, but I thought I keep it this way just in case. But considering @erikerikson use case, I think we don't need that anymore.
